### DISCUSSION TITLE
Expo lockscreen controls

### DISCUSSION THREAD

**Discussion by carvalhochris** [expo/expo#28068] [Proposals and Ideas]
This is very much the only missing piece when building a music player in Expo atm.

I have audio playing in the background, but no lockscreen controls.

Would love to help make this happen in any way I can.

**Comment by sebnun** [comment]
I see there is a new alpha `expo-audio` package, but looks like still there is no lockscreen controls functionality.
https://docs.expo.dev/versions/v52.0.0/sdk/audio/

 Is there any plan to add lockscreen controls before the stable release?

**Comment by tsapeta** [comment]
Yes, there is, it's on our roadmap. `expo-video` already does support lockscreen controls, `expo-audio` is just a bit behind in terms of the functionalities.
Since this feature is going to be quite similar in both packages, there is also a discussion about making a separate package for this: https://github.com/expo/expo/discussions/31809

**Comment by trentcowden** [comment]
@tsapeta Is there any update on this? Just installed Expo SDK 53 and noticed there's still no lockscreen controls. This would be a great feature to have for apps that use background audio.

**Comment by jwhscholten** [comment]
I used expo-video with some additional audio config. Works pretty ok on iOS and Android.

```
import React, { useState, useEffect, useMemo } from "react";
import {
  View,
  TouchableOpacity,
  Text as RNText,
  StyleSheet,
} from "react-native";
import { useVideoPlayer } from "expo-video";
import { setAudioModeAsync } from "expo-audio";
import Icon from "@expo/vector-icons/FontAwesome";
import { MaterialIcons } from "@expo/vector-icons";

// Dummy data (replace with your own as needed)
const DUMMY_SOURCE = { uri: "https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3" };
const DUMMY_TITLE = "Sample Audio";
const DUMMY_ARTWORK = "https://i.imgur.com/TKyCtYD.png";
const PRIMARY_COLOR = "#3366FF";

export default function TestAudioPlayerAV() {
  // Enrich source with metadata
  const sourceWithMetadata = useMemo(() => ({
    ...DUMMY_SOURCE,
    metadata: {
      title: DUMMY_TITLE,
      artwork: DUMMY_ARTWORK,
    },
  }), []);

  // Initialize the VideoPlayer (audio-only)
  const player = useVideoPlayer(sourceWithMetadata, (p) => {
    p.loop = false;
    p.staysActiveInBackground = true;
    p.showNowPlayingNotification = true;
    p.audioMixingMode = "doNotMix";
  });

  const [isPlaying, setIsPlaying] = useState(false);
  const [duration, setDuration] = useState(0); // ms
  const [position, setPosition] = useState(0); // ms

  // Poll for playback progress
  useEffect(() => {
    const interval = setInterval(() => {
      if (!player) return;
      setPosition((player.currentTime ?? 0) * 1000);
      setDuration((player.duration ?? 0) * 1000);
    }, 500);
    return () => clearInterval(interval);
  }, [player.id]);

  // Configure audio mode on mount
  useEffect(() => {
    setAudioModeAsync({
      shouldPlayInBackground: true,
      playsInSilentMode: true,
      interruptionMode: "doNotMix",
      interruptionModeAndroid: "doNotMix",
      allowsRecording: false,
      shouldRouteThroughEarpiece: false,
    }).catch(console.error);
  }, []);

  // Replace player source if metadata changes
  useEffect(() => {
    if (sourceWithMetadata) {
      player.replace(sourceWithMetadata, true).catch(console.error);
    }
  }, [JSON.stringify(sourceWithMetadata)]);

  // Playback controls
  const togglePlayback = async () => {
    if (!player) return;
    try {
      if (isPlaying) {
        await player.pause();
        setIsPlaying(false);
      } else {
        await player.play();
        setIsPlaying(true);
      }
    } catch (err) {
      console.error("Playback error:", err);
    }
  };

  const restart = async () => {
    if (player) {
      try {
        await player.replay();
        setIsPlaying(true);
      } catch (err) {
        console.error("Restart error:", err);
      }
    }
  };

  const skipForward = async () => {
    if (!player) return;
    try {
      const current = player.currentTime ?? 0;
      const newTime = Math.min(current + 10, player.duration ?? current + 10);
      if (player.seekTo) {
        await player.seekTo(newTime);
      } else {
        player.currentTime = newTime;
      }
    } catch (err) {
      console.error("Skip forward error:", err);
    }
  };

  const skipBackward = async () => {
    if (!player) return;
    try {
      const current = player.currentTime ?? 0;
      const newTime = Math.max(current - 10, 0);
      if (player.seekTo) {
        await player.seekTo(newTime);
      } else {
        player.currentTime = newTime;
      }
    } catch (err) {
      console.error("Skip backward error:", err);
    }
  };

  const formatTime = (millis) => {
    if (!millis || millis < 0) return "00:00";
    const totalSeconds = Math.floor(millis / 1000);
    const minutes = Math.floor(totalSeconds / 60);
    const seconds = totalSeconds % 60;
    return `${minutes}:${String(seconds).padStart(2, "0")}`;
  };

  const progress = duration > 0 ? position / duration : 0;

  return (
    <View style={styles.container}>
      {/* Audio title */}
      <View style={styles.titleContainer}>
        <RNText style={styles.title}>{DUMMY_TITLE}</RNText>
      </View>

      {/* Progress bar */}
      <View style={styles.progressSection}>
        <View style={styles.progressBarContainer}>
          <View style={styles.progressTrack}>
            <View
              style={[
                styles.progressFill,
                {
                  width: `${progress * 100}%`,
                  backgroundColor: PRIMARY_COLOR,
                },
              ]}
            />
            <View
              style={[
                styles.progressThumb,
                {
                  left: `${progress * 100}%`,
                  backgroundColor: PRIMARY_COLOR,
                },
              ]}
            />
          </View>
        </View>
        <View style={styles.timeContainer}>
          <RNText style={styles.timeText}>{formatTime(position)}</RNText>
          <RNText style={styles.timeText}>{formatTime(duration)}</RNText>
        </View>
      </View>

      {/* Controls */}
      <View style={styles.controls}>
        <TouchableOpacity style={styles.circleButton} onPress={restart}>
          <Icon name="refresh" size={18} color={PRIMARY_COLOR} />
        </TouchableOpacity>

        <TouchableOpacity style={styles.skipButton} onPress={skipBackward}>
          <MaterialIcons
            name="replay-10"
            size={28}
            color={PRIMARY_COLOR}
          />
        </TouchableOpacity>

        <TouchableOpacity
          style={[
            styles.playButton,
            { backgroundColor: PRIMARY_COLOR },
          ]}
          onPress={togglePlayback}
        >
          <Icon
            name={isPlaying ? "pause" : "play"}
            size={24}
            color="white"
            style={isPlaying ? styles.pauseIcon : styles.playIcon}
          />
        </TouchableOpacity>

        <TouchableOpacity style={styles.skipButton} onPress={skipForward}>
          <MaterialIcons
            name="forward-10"
            size={28}
            color={PRIMARY_COLOR}
          />
        </TouchableOpacity>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    alignItems: "center",
    paddingVertical: 24,
    width: "100%",
  },
  titleContainer: {
    alignItems: "center",
    marginBottom: 24,
  },
  title: {
    fontSize: 16,
    fontWeight: "600",
  },
  controls: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center",
    width: "100%",
  },
  circleButton: {
    width: 36,
    height: 36,
    borderRadius: 18,
    justifyContent: "center",
    alignItems: "center",
    marginHorizontal: 12,
  },
  skipButton: {
    width: 40,
    height: 40,
    justifyContent: "center",
    alignItems: "center",
    marginHorizontal: 16,
  },
  playButton: {
    width: 64,
    height: 64,
    borderRadius: 32,
    justifyContent: "center",
    alignItems: "center",
    marginHorizontal: 16,
  },
  progressSection: {
    width: "100%",
    marginBottom: 24,
  },
  progressBarContainer: {
    width: "100%",
    height: 20,
    justifyContent: "center",
  },
  progressTrack: {
    width: "100%",
    height: 4,
    backgroundColor: "#E5E5E5",
    borderRadius: 2,
    position: "relative",
  },
  progressFill: {
    height: "100%",
    borderRadius: 2,
    position: "absolute",
  },
  progressThumb: {
    position: "absolute",
    width: 16,
    height: 16,
    borderRadius: 8,
    top: -6,
    marginLeft: -8,
    elevation: 2,
    shadowColor: "#000",
    shadowOffset: { width: 0, height: 1 },
    shadowOpacity: 0.2,
    shadowRadius: 1,
  },
  timeContainer: {
    flexDirection: "row",
    justifyContent: "space-between",
    marginTop: 8,
    width: "100%",
  },
  timeText: {
    fontSize: 14,
    color: "#777777",
  },
  playIcon: {
    marginLeft: 4,
  },
  pauseIcon: {
    marginLeft: 0,
  },
});
```

**Comment by swkovma** [comment]
@tsapeta Is there any update on this?

**Comment by alanjhughes** [comment]
Hi @swkovma - there are two draft PRs to add support for this. It will be available in sdk 54

**Reply by arjunkyal** [reply]
Hey is the fix out?

**Reply by swkovma** [reply]
Hi @alanjhughes 

Can you please confirm the issue is fixed?

**Reply by alanjhughes** [reply]
This was not an issue, it's a feature request. It did not make it in to sdk 54

**Reply by Tanmay-97** [reply]
@alanjhughes any chances this may release soon? totally require it for my project

**Comment by dani-z** [comment]
Not that I want to interfere with the plans @alanjhughes has for expo-audio, for those who want to have the lock screen controls I have found https://rnap.dev/ and works perfectly for that.

**Comment by luklapp** [comment]
@tsapeta According to [the docs](https://docs.expo.dev/versions/v54.0.0/sdk/audio/#setactiveforlockscreenactive-metadata-options),  there are some methods available, like `setActiveForLockScreen` or `updateLockScreenMetadata`. I installed the latest version (1.0.14, SDK 54), but it looks like the methods are not available yet 🤔

**Reply by ardaydilek** [reply]
Yep. cannot see them in the AudioModule.types.ts in the latest SDK

**Reply by Tanmay-97** [reply]
Yeah, not seeing them. Hope this releases soon!

**Comment by IceDev-xyz** [comment]
After fighting for 1hr on this, I made it work (Claude did) with the SDK 54..

1. Update to expo-audio 1.1.0+
`"expo-audio": "~1.1.0"`
2. Add iOS background mode in app.json
```
{
  "expo": {
    "ios": {
      "infoPlist": {
        "UIBackgroundModes": ["audio"]
      }
    }
  }
}
```
3. Configure audio mode with "doNotMix" (this was key!)
```
import { setAudioModeAsync } from "expo-audio";

await setAudioModeAsync({
  playsInSilentMode: true,
  shouldPlayInBackground: true,
  interruptionMode: "doNotMix",  // ← Must be "doNotMix", NOT "duckOthers"
  interruptionModeAndroid: "doNotMix",
});
```
4. Call setActiveForLockScreen AFTER playback starts with a small delay

```
// After calling player.play(), wait a moment for the audio session to activate
setTimeout(() => {
  player.setActiveForLockScreen(
    true,
    {
      title: "Track Title",
      artist: "Artist Name",
      artworkUrl: "https://example.com/artwork.jpg",
    },
    {
      showSeekForward: true,
      showSeekBackward: true,
    }
  );
}, 500);
```

**Reply by kadikraman** [reply]
Thanks for calling out about `interruptionMode` needing to be set to `"doNotMix"` - just merged a change to add a note about it to the docs https://github.com/expo/expo/pull/41635

As @davidfloegel confirmed below, the `setTimeout` should not be necessary since you can activate the controls at any time - the player doesn't need to be playing. Could you try again without the timeout? Could be one of those debugging sessions where lots of different workarounds were tried, but in the end not all changes were actually needed.

**Reply by oojr** [reply]
timeout doesn't work I'm on expo/expo-audio 55, react-native 83

**Reply by davidfloegel** [reply]
> timeout doesn't work I'm on expo/expo-audio 55, react-native 83

@oojr a snippet of your code would be helpful! As mentioned, `setTimeout` is _not actually necessary_ - what is necessary is setting the `interruptionMode` to `doNotMix`

**Comment by davidfloegel** [comment]
@IceDev-xyz Great catch re the `interruptionMode`, I managed to get it to working using `doNotMix`.

I didn't need the timeout though.

The Expo team ( @kadikraman ) is kindly updating the docs to include this.

**Comment by davidfloegel** [comment]
Hi Expo Audio team 👋

We’ve been integrating expo-audio into a Redux-driven playback experience (the core of our app, serving a million music super fans) and wanted to share a couple of API pain points that are difficult to solve cleanly with the current event surface.

## Our setup

We wrap `expo-audio` in an AudioPlayer singleton that abstracts player lifecycle + platform audio mode configuration, and exposes simple methods (play, pause, resume, seekTo, stop, updateMetaData). Internally, this singleton subscribes to:

```js
player.addListener("playbackStatusUpdate", (status) => this._handleStatus(status))
```

…and translates status changes into Redux actions (`playbackPaused`, `playbackResumed`, `playbackInfoUpdated`, `trackEnded`, etc.). This structure is important for us because Redux is our single source of truth for the UI and for coordinating playback state across screens.

## Challenges we’re currently facing

### 1) Reliably detecting “track ended” in playbackStatusUpdate

We dispatch trackEnded when:
```js
if (status.isLoaded && status.didJustFinish) { ... }
```

However, on Android we’re seeing an issue where didJustFinish can be stale and may fire multiple times across successive status updates. To prevent race conditions / duplicate trackEnded dispatches, we’ve had to add our own debounce and deduping:

1. Track lastFinishTime and ignore finishes within a short window (we use 500ms)
2. Track lastFinishedTrackId so the same track doesn’t “finish” twice
3. Maintain currentTrackId inside our singleton because the status event itself doesn’t reliably give us an app-level track identity

This makes a seemingly simple “track finished” signal surprisingly tricky and forces custom code to do defensive bookkeeping.

### 2) Detecting lock screen pause/resume is extremely difficult

We also need to know when playback was paused/resumed via the lock screen / external media controls, because we keep canonical playback state in Redux (a play/pause button for example) and want to reflect user intent accurately.

Right now, the only way to infer this is via manual conditions in `playbackStatusUpdate`, comparing our last known `isPlaying` state against the incoming status. In our implementation we effectively do:

1. Compute `isPlayingOrReady = status.playing || status.playbackState === "readyToPlay"`
2. Track internal flags like `isPlaying`
3. Add extra state like `isSwitchingTrack` because track replacement (replace() + play()) can look like a pause/resume transition in the status stream

So we end up with logic like:

1. “We were playing, now we’re not playing/ready, and we’re not switching tracks → infer lock screen pause”
2. “We were not playing, now we are playing/ready → infer lock screen resume”

This works _most of the time_, but it’s fragile and grows in complexity as more controls come into play (seek forward/backward, skip forward/backward, etc.). It also makes it hard to avoid edge cases like double triggers or ambiguous transitions.

## Proposal: dedicated listeners for system/lock screen control actions

To make these scenarios robust and reduce application-side inference, it would be hugely helpful if expo-audio exposed explicit control/action listeners for media control intent, e.g.:

```js
// Example sources
type Source = 'lockscreen' | 'in-app' | 'system'

addListener("playbackPaused", (source: Source) => {})
addListener("playbackResumed", (source: Source) => {})
addListener("seekedForward", (source: Source) => {})
addListener("seekedBackward", (source: Source) => {})
addListener("skippedForward", (source: Source) => {})
addListener("skippedBackward", (source: Source) => {})

// this would also be nice:
addListener("trackEnded", () => {})
```

This would let apps:

1. Reliably distinguish user intent (lock screen vs in-app)
2. Keep Redux (or any external store) in sync without heuristic comparisons
3. Avoid custom “switching track” flags and other defensive logic
4. Scale cleanly as more system controls are added

Having first-class, intent-based signals would remove a lot of brittle app code and improve reliability across platforms - especially Android. It would also put `expo-audio` ahead of other react-native audio libraries that have similar functionality but most of the time it's still brittle and unreliable.

Happy to share more details or help test if this direction is interesting. Thanks again for all the work on expo-audio 🙏

**Comment by aayla-secura** [comment]
Am I missing something or is there currently no way to use lock screen controls AND playlist? It seems like you either use `AudioPlayer` which exposes lock screen controls but doesn't support queue with gapless transitioning, _or_ you use `AudioPlaylist` which doesn't export controls for the lock screen or the underlying `AudioPlayer`...
